api/src/main/java/org/infinispan/api/async/AsyncCache.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.api.async;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.api.common.CacheEntry;
+import org.infinispan.api.common.Flag;
+import org.infinispan.api.common.events.cache.CacheContinuousQueryEvent;
+import org.infinispan.api.common.events.cache.CacheEntryEvent;
+import org.infinispan.api.common.events.cache.CacheEntryEventType;
+import org.infinispan.api.common.events.cache.CacheListenerOptions;
+import org.infinispan.api.common.tasks.EntryConsumerTask;
+import org.infinispan.api.configuration.CacheConfiguration;
+
+/**
+ * @since 14.0
+ **/
+public interface AsyncCache<K, V> {
+   /**
+    * The name of the cache.
+    *
+    * @return
+    */
+   String name();
+
+   /**
+    * The configuration for this cache.
+    *
+    * @return
+    */
+   CompletionStage<CacheConfiguration> configuration();
+
+   /**
+    * Return the container of this cache
+    * @return
+    */
+   AsyncContainer container();
+
+   /**
+    * Returns a new instance of this cache with the supplied {@link Flag}s
+    *
+    * @param flags
+    * @return
+    */
+   AsyncCache<K, V> withFlags(Set<Flag> flags);
```

Comment:
Are there Flags that you have in mind for the new API?

Flag creep was a problem with the old API. Where possible we should try to define the API so that additional flags aren't required.

Reply:
I've removed all traces of flags